lib/webview-helpers.js

```diff
@@ -12,48 +13,97 @@ const CROSSWALK_SOCKET_SUFFIX = '_devtools_remote';
     line = line.trim();
+    let regexMatch;
+    if ((regexMatch = (line.match(WEBVIEW_REGEXP) || line.match(CROSSWALK_REGEXP)))) {
+      procs.push(regexMatch[0]);
+    }
+  }
 
-    if (deviceSocket) {
-      if (line.indexOf(`@${deviceSocket}`) === line.length - deviceSocket.length - 1) {
-        if (deviceSocket === 'chrome_devtools_remote') {
-          webviews.push(CHROMIUM_WIN);
-          continue;
-        }
-      }
+  // sometimes the webview process shows up multiple times per app
+  return _.uniq(procs);
+}
+
+// This function retrieves a list of system processes that look like webviews,
+// and returns them along with the webview context name appropriate for it.
+// If we pass in a deviceSocket, we only attempt to find webviews which match
+// that socket name (this is for apps which embed Chromium, which isn't the
+// same as chrome-backed webviews). Returns a list of objects of the form
+// {proc: <proc-name>, webview: <webview-name>}
+// TODO: some of this function probably belongs in appium-adb?
+async function webviewsFromProcs (adb, deviceSocket) {
+  const procs = await getPotentialWebviewProcs(adb);
+  const webviews = [];
+  for (const proc of procs) {
+    if (deviceSocket === 'chrome_devtools_remote' && proc === `@${deviceSocket}`) {
+      webviews.push({proc, webview: CHROMIUM_WIN});
+      continue;
     }
 
     let webviewPid;
     let crosswalkWebviewSocket;
-    if ((webviewPid = line.match(WEBVIEW_REGEXP))) {
+    if ((webviewPid = proc.match(WEBVIEW_REGEXP))) {
       // for multiple webviews a list of 'WEBVIEW_<index>' will be returned
       // where <index> is zero based (same is in selendroid)
-      webviews.push(`${WEBVIEW_BASE}${webviewPid[1]}`);
-    } else if ((crosswalkWebviewSocket = line.match(CROSSWALK_REGEXP))) {
+      webviews.push({proc, webview: `${WEBVIEW_BASE}${webviewPid[1]}`});
+    } else if ((crosswalkWebviewSocket = proc.match(CROSSWALK_REGEXP))) {
       if (deviceSocket) {
-        if (crosswalkWebviewSocket[0].slice(1) === deviceSocket) {
-          webviews.push(`${WEBVIEW_BASE}${crosswalkWebviewSocket[1]}`);
+        if (crosswalkWebviewSocket[0] === `@${deviceSocket}`) {
+          webviews.push({proc, webview: `${WEBVIEW_BASE}${crosswalkWebviewSocket[1]}`});
         }
       } else {
-        webviews.push(`${WEBVIEW_BASE}${crosswalkWebviewSocket[1]}${CROSSWALK_SOCKET_SUFFIX}`);
+        webviews.push({proc, webview: `${WEBVIEW_BASE}${crosswalkWebviewSocket[1]}${CROSSWALK_SOCKET_SUFFIX}`});
       }
     }
   }
-  return _.uniq(webviews);
+  return webviews;
+}
+
+// given a 'webview-proc' object, check whether we can detect any active pages
+// corresponding to that webview. This is used by getWebviews to filter out any
+// webview-looking processes which have a remote debug port open but which
+// aren't actually running any pages, because such processes can't be automated
+// by chromedriver anyway.
+async function webviewHasPages (adb, wp, webviewDevtoolsPort) {
```

Comment:
`wp` could be destructed rather directly here or in the function body to enhance readability